src/community/schemaless-features/schemaless-core/src/main/java/org/geoserver/schemalessfeatures/type/SchemalessComplexTypeImpl.java

```diff
@@ -0,0 +1,127 @@
+/* (c) 2021 Open Source Geospatial Foundation - all rights reserved
+ * This code is licensed under the GPL 2.0 license, available at the root
+ * application directory.
+ */
+package org.geoserver.schemalessfeatures.type;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.geotools.feature.NameImpl;
+import org.geotools.feature.type.AttributeTypeImpl;
+import org.opengis.feature.Property;
+import org.opengis.feature.type.AttributeType;
+import org.opengis.feature.type.Name;
+import org.opengis.feature.type.PropertyDescriptor;
+import org.opengis.filter.Filter;
+import org.opengis.util.InternationalString;
+
+/** Concrete implementation of a SchemalessComplexType */
+public class SchemalessComplexTypeImpl extends AttributeTypeImpl implements SchemalessComplexType {
+
+    private final Collection<PropertyDescriptor> properties;
+
+    private final Map<Name, PropertyDescriptor> propertyMap;
+
+    public SchemalessComplexTypeImpl(
+            Name name,
+            Collection<PropertyDescriptor> properties,
+            boolean identified,
+            boolean isAbstract,
+            List<Filter> restrictions,
+            AttributeType superType,
+            InternationalString description) {
+        super(name, Collection.class, identified, isAbstract, restrictions, superType, description);
+        Map<Name, PropertyDescriptor> localPropertyMap;
+        if (properties == null) {
+            localPropertyMap = new HashMap<>();
+        } else {
+            localPropertyMap = new HashMap<>();
+            for (PropertyDescriptor pd : properties) {
+                if (pd == null) {
+                    // descriptor entry may be null if a request was made for a property that does
+                    // not exist
+                    throw new NullPointerException(
+                            "PropertyDescriptor is null - did you request a property that does not exist?");
+                }
+                localPropertyMap.put(pd.getName(), pd);
+            }
+        }
+        this.properties = properties;
+        this.propertyMap = localPropertyMap;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public Class<Collection<Property>> getBinding() {
```

Comment:
The Class is not extending ComplexTypeImpl but AttributeTypeImpl thus those method need to be implemented. This has been done because the descriptors collection is being made unmodifyiable in the constructor and has private access. To be able to add new descriptor to it I would have need to assign them to a new attribute with a normal Collection, or change directly the ComplexTypeImpl class. Seemed to me a more straight forward solution to implement only the interface and extending the AttributeTyeImpl.